File: 2025/01 January/20 Merge two sorted linked lists/main.cpp
// Link: https://www.geeksforgeeks.org/problems/merge-two-sorted-linked-lists/1

#include <iostream>
using namespace std;

struct Node
{
    int data;
    struct Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

void printList(struct Node *head)
{
    struct Node *temp = head;
    while (temp != NULL)
    {
        cout << temp->data << ' ';
        temp = temp->next;
    }
    cout << "\n~\n";
}

Node *insertSorted(Node *head, int data)
{
    Node *new_node = new Node(data);
    if (!head || head->data >= data)
    {
        new_node->next = head;
        return new_node;
    }

    Node *current = head;
    while (current->next && current->next->data < data)
    {
        current = current->next;
    }
    new_node->next = current->next;
    current->next = new_node;
    return head;
}

Node *createLinkedList(int arr[], int n)
{
    Node *head = NULL;
    for (int i = 0; i < n; i++)
    {
        head = insertSorted(head, arr[i]);
    }
    return head;
}

class Solution
{
public:
    Node *sortedMerge(Node *head1, Node *head2)
    {
        Node *head = (head1->data <= head2->data) ? head1 : head2;
        Node *tail = head;
        Node *a = head1;
        Node *b = head2;
        (head1->data <= head2->data) ? a = a->next : b = b->next;
        while (a != NULL && b != NULL)
        {
            tail->next = (a->data <= b->data) ? a : b;
            (a->data <= b->data) ? a = a->next : b = b->next;
            tail = tail->next;
        }
        tail->next = (b == NULL) ? a : b;
        return head;
    }
};

int main()
{
    Solution object;
    int arr11[] = {5, 10, 15, 40};
    int arr12[] = {2, 3, 20};
    int n1 = sizeof(arr11) / sizeof(arr11[0]);
    int n2 = sizeof(arr12) / sizeof(arr12[0]);
    Node *head1 = createLinkedList(arr11, n1);
    Node *head2 = createLinkedList(arr12, n2);
    Node *merge = object.sortedMerge(head1, head2);
    printList(merge);

    int arr21[] = {1, 1};
    int arr22[] = {2, 4};
    n1 = sizeof(arr21) / sizeof(arr21[0]);
    n2 = sizeof(arr22) / sizeof(arr22[0]);
    head1 = createLinkedList(arr21, n1);
    head2 = createLinkedList(arr22, n2);
    merge = object.sortedMerge(head1, head2);
    printList(merge);

    return 0;
}

File: 2025/01 January/20 Merge two sorted linked lists/main.py
# Link: https://www.geeksforgeeks.org/problems/merge-two-sorted-linked-lists/1


class Node:

    def __init__(self, data):
        self.data = data
        self.next = None


def print_list(head):
    temp = head
    while temp:
        print(temp.data, end=" ")
        temp = temp.next
    print()
    print("~")


def insert_sorted(head, data):
    new_node = Node(data)
    if not head or head.data >= data:
        new_node.next = head
        return new_node

    current = head
    while current.next and current.next.data < data:
        current = current.next

    new_node.next = current.next
    current.next = new_node

    return head


class LinkedList:

    def __init__(self):
        self.head = None

    def insert(self, data):
        self.head = insert_sorted(self.head, data)


def createLinkedList(arr):
    linked_list = LinkedList()
    for i in arr:
        linked_list.insert(i)
    return linked_list.head


class Solution:
    def sortedMerge(self, head1, head2):
        temp = Node(0)
        head = temp
        while head1 and head2:
            if head1.data <= head2.data:
                head.next = head1
                head1 = head1.next
            else:
                head.next = head2
                head2 = head2.next
            head = head.next
        if head1:
            head.next = head1
        else:
            head.next = head2
        return temp.next


if __name__ == "__main__":
    object = Solution()
    arr11 = [5, 10, 15, 40]
    arr12 = [2, 3, 20]
    head1 = createLinkedList(arr11)
    head2 = createLinkedList(arr12)
    merge = object.sortedMerge(head1, head2)
    print_list(merge)

    arr21 = [1, 1]
    arr22 = [2, 4]
    head1 = createLinkedList(arr21)
    head2 = createLinkedList(arr22)
    merge = object.sortedMerge(head1, head2)
    print_list(merge)
